Add smarter getter for block slipperiness

howtonotwin opened the pull request:
Allow a block access to the blockstate, world, position, and entity when getting slipperiness with new `Block::getSlipperiness` method. Change usages of `Block::slipperiness` to `Block::getSlipperiness` in vanilla. Deprecate old field.

[Boat Blaster](https://gist.github.com/howtonotwin/3ff1260ac20d6befbf0784708e30c14f) test mod gist. As per the name, it adds a block (`boat_blaster:boat_blaster`) that blasts boats at high speed (think a hundred blocks/tick) by using the new getter to give the block a slipperiness of `2` for `EntityBoat`s. It retains the default for other entities. (Don't actually be in the boat when this happens. MC complains about the high speed and everything goes choppy and weird. Works fine as long as anything *but* a player is in the boat.)

Resolves #1342.

howtonotwin committed:
Add smarter getter for block slipperiness

Allow a block access to the blockstate, world, position, and entity when
getting slipperiness with new `Block::getSlipperiness` method.
Change usages of `Block::slipperiness` to `Block::getSlipperiness` in
vanilla.
Deprecate old field.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=3620) <br/>All committers have signed the CLA.

liach commented:
You need to fix the merge conflict.

bs2609 commented:
This needs updating again to resolve conflicts @howtonotwin

howtonotwin commented:
Is there a good, tested way to fix patch conflicts? The way I did it was

1. `git reset --hard HEAD`
2. Checkout `slipperiness` branch
3. `./gradlew setupForge`
4. Copy `projects/Forge/src/main/java` into external location
5. Repeat for `1.11.x` branch and for `git merge-base 1.11.x slipperiness` commit
6. `for file in $(find base -type f -printf '%P\n'); do git merge-file slipperiness/$file base/$file upstream/$file; done`
7. Copy resulting sources back into Forge tree
8. `./gradlew genPatches`

But I'm not sure if this will produce any errors in other cases, or if there's an easier way. I checked the output just in case and it seems fine.

mezz commented:
No, fixing patch conflicts is not particularly easy. I think you have the right idea. I tend to just fix the code manually and then `genPatches`. This one just happens to touch a pretty popular file, `Block.java`, so it probably conflicts more often than other PRs.

liach commented on patches/minecraft/net/minecraft/block/Block.java.patch:
```diff
@@ -207,6 +215,23 @@
++    /**
++     * Gets the slipperiness at the given location at the given state. Normally between 0 and 1.
++     * <p>
++     * Note that entities may reduce slipperiness by a certain factor of their own; for {@link net.minecraft.entity.EntityLivingBase}, this is {@code .91}. {@link net.minecraft.entity.item.EntityItem} uses {@code .98}, and {@link net.minecraft.entity.projectile.EntityFishHook} uses {@code .92}.
```
Can you split this in multiple lines?

howtonotwin commented:
Should there be a `setDeaultSlipperiness(Float)` so that someone can set the field without generating a deprecation warning? It'll come with a Javadoc message that signifies that actual slipperiness comes from `getSlipperiness` and the set value may not be accurate, or even be wholly ignored.

howtonotwin committed:
Line wrap Block::getSlipperiness Javadoc

bs2609 commented on patches/minecraft/net/minecraft/block/Block.java.patch:
```diff
@@ -207,6 +215,25 @@
++     * <p>
++     * Note that entities may reduce slipperiness by a certain factor of their own; for {@link net.minecraft.entity.EntityLivingBase},
++     * this is {@code .91}. {@link net.minecraft.entity.item.EntityItem} uses {@code .98}, and
++     * {@link net.minecraft.entity.projectile.EntityFishHook} uses {@code .92}.
```
Could you split this so each example gets its own line?

Like:
```
 * Note that entities may reduce slipperiness by a certain factor of their own;
 * for {@link net.minecraft.entity.EntityLivingBase}, this is {@code .91},
 * {@link net.minecraft.entity.item.EntityItem} uses {@code .98},
 * and {@link net.minecraft.entity.projectile.EntityFishHook} uses {@code .92}.
```

howtonotwin replied:
How about a table?

howtonotwin committed:
Re-line-wrap Block::getSlipperiness Javadoc

kashike commented:
> Should there be a `setDeaultSlipperiness(Float)` so that someone can set the field without generating a deprecation warning? It'll come with a Javadoc message that signifies that actual slipperiness comes from `getSlipperiness` and the set value may not be accurate, or even be wholly ignored.

Wouldn't hurt, and would prevent complaints about deprecation warnings.

howtonotwin committed:
Add Block::setDefaultSlipperiness

mezz commented on patches/minecraft/net/minecraft/block/Block.java.patch:
```diff
@@ -207,6 +215,42 @@
++     * @param slipperiness the base slipperiness of this block
++     * @see #getSlipperiness(IBlockState, IBlockAccess, BlockPos, Entity)
++     */
++    public void setDefaultSlipperiness(float slipperiness) {
```
formatting: braces must be on a new line

mezz reviewed (changes requested):
Looks good. 

Please fix formatting I commented on.

Also please include the `BoatBlaster` test mod as `SlipperinessTest` in this PR. It should go with the other forge test mods, and be disabled by default.

howtonotwin committed:
Fix brace on Block::setDefaultSlipperiness

howtonotwin committed:
Add "Slippperiness Test" mod

Create a "bloat_blaster" block that exponentially accelerates boats`

howtonotwin committed:
Merge branch '1.11.x' into slipperiness

howtonotwin commented:
Should this be rebased to `1.12.x`?

mezz commented:
Yes, this should target 1.12.x.
If you want to also target 1.11.x, make a new 1.12.x PR and leave this one open.

howtonotwin committed:
Fix a typo

(cherry picked from commit 3a23c1b2ef9bba0abb63d590953e177c1eaa1b89)

mezz commented:
If you want, please update this to match the 1.12 version and I can merge it. Otherwise we should close it.

howtonotwin committed:
Fix MC-119811

Patch `EntityBoat` to instantly complete a pending movement lerp when
something can steer the boat.

(cherry picked from commit ed5b5e1c7aa22f5216937ba32eb66aac3eb27b71)

howtonotwin commented:
That should be it.
